searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.cpp

```diff
@@ -303,6 +351,101 @@ PutTask::run()
+    FieldValue::SP _field_value;
+    std::promise<std::unique_ptr<PrepareResult>> _result_promise;
+
+public:
+    PreparePutTask(SerialNum serial_num_in,
+                   uint32_t docid_in,
+                   const AttributeWriter::WriteField& field,
+                   std::shared_ptr<DocumentFieldExtractor> field_extractor);
+    ~PreparePutTask() override;
+    void run() override;
+    SerialNum serial_num() const { return _serial_num; }
+    uint32_t docid() const { return _docid; }
+    AttributeVector& attr() { return _attr; }
+    FieldValue::SP field_value() { return _field_value; }
+    std::future<std::unique_ptr<PrepareResult>> result_future() {
+        return _result_promise.get_future();
+    }
+};
+
+PreparePutTask::PreparePutTask(SerialNum serial_num_in,
+                               uint32_t docid_in,
+                               const AttributeWriter::WriteField& field,
+                               std::shared_ptr<DocumentFieldExtractor> field_extractor)
+    : _serial_num(serial_num_in),
+      _docid(docid_in),
+      _attr(field.getAttribute()),
+      _field_value(),
+      _result_promise()
+{
+    // Note: No need to store the field extractor as we are not extracting struct fields.
+    auto value = field_extractor->getFieldValue(field.getFieldPath());
+    _field_value.reset(value.release());
+}
+
+PreparePutTask::~PreparePutTask() = default;
+
+void
+PreparePutTask::run()
+{
+    if (_attr.getStatus().getLastSyncToken() < _serial_num) {
+        if (_field_value.get()) {
+            _result_promise.set_value(AttributeUpdater::prepare_set_value(_attr, _docid, *_field_value));
+        }
+    }
+}
+
+class CompletePutTask : public vespalib::Executor::Task {
+private:
+    const SerialNum _serial_num;
+    const uint32_t _docid;
+    AttributeVector& _attr;
+    FieldValue::SP _field_value;
+    std::future<std::unique_ptr<PrepareResult>> _result_future;
+    const bool _immediate_commit;
+    std::remove_reference_t<AttributeWriter::OnWriteDoneType> _on_write_done;
+
+public:
+    CompletePutTask(PreparePutTask& prepare_task,
+                    bool immediate_commit,
+                    AttributeWriter::OnWriteDoneType on_write_done);
+    ~CompletePutTask() override;
+    void run() override;
+};
+
+CompletePutTask::CompletePutTask(PreparePutTask& prepare_task,
+                                 bool immediate_commit,
+                                 AttributeWriter::OnWriteDoneType on_write_done)
+    : _serial_num(prepare_task.serial_num()),
+      _docid(prepare_task.docid()),
+      _attr(prepare_task.attr()),
+      _field_value(prepare_task.field_value()),
+      _result_future(prepare_task.result_future()),
+      _immediate_commit(immediate_commit),
+      _on_write_done(on_write_done)
+{
+}
+
+CompletePutTask::~CompletePutTask() = default;
+
+void
+CompletePutTask::run()
+{
+    if (_attr.getStatus().getLastSyncToken() < _serial_num) {
+        auto result = _result_future.get();
```

Comment:
This will hang if promise is never set, e.g. when PreparePutTask::run() doesn't set promise due to field value not being set.

Reply:
Fixed